### PR TITLE
[pekko-testkit-typed] Support for JUnit5 in testkit 

### PR DESCRIPTION
**Pull request by thmue**
Due to some issues with the scala setup i created a new PR, which is based on the current main branch. 

This PR replaces https://github.com/apache/incubator-pekko/pull/445

The extension is activated through an annotation:
@ExtendWith(TestKitJUnit5Extension.class)
on the class level.

Additionally, a testKit with a @Junit5TestKit annotation needs to be created.
A builder pattern is used to construct the Testkit instance.

@JUnit5TestKit
public ActorTestKit testKit = new JUnit5TestKitBuilder().build();

The extension itself uses the JUnit5 AfterAllCallback to shutdown the testkit after all tests are executed.

### CODE REVIEW

File: actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 public class LogCapturingExampleTest {
 
-  @ClassRule public static final TestKitJunitResource testKit = new TestKitJunitResource();
+  @ClassRule public static final TestKitJUnitResource testKit = new TestKitJUnitResource();
```
Comment (pjfanning) [inline actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java:29]: Can we leave the legacy Junit 4 classes alone?

I must also admit to preferring that the new classes are `Junit5` to match the pre-existing standard. I am going to -1 the PR due to the renaming of the existing classes.

**Review by pjfanning — CHANGES_REQUESTED** [review]

File: actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java (hunk 1 above)
Comment (thmue) [reply actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java]: ok i changed it back :-)

File: actor-testkit-typed/src/main/java/org/apache/pekko/actor/testkit/typed/annotations/JUnit5TestKit.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package org.apache.pekko.actor.testkit.typed.annotations;
+
+import java.lang.annotation.*;
+
+@Documented
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface JUnit5TestKit {}
```
Comment (pjfanning) [inline actor-testkit-typed/src/main/java/org/apache/pekko/actor/testkit/typed/annotations/JUnit5TestKit.java:25]: * thanks for reverting the changes to the junit4 classes
* with the junit5 classes, can we be consistent - it's either `JUnIt` or it's `Junit` but we in this PR, there is a mix.

Comment (thmue) [reply actor-testkit-typed/src/main/java/org/apache/pekko/actor/testkit/typed/annotations/JUnit5TestKit.java]: Thanks for spotting this inconsistency, fixed

File: actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java (hunk 1 above)
Comment (mdedetrich) [reply actor-testkit-typed/src/test/java/jdocs/org/apache/pekko/actor/testkit/typed/javadsl/LogCapturingExampleTest.java]: I disagree, I think the class should be named `JUnit5` and we should have a deprecated alias for `Junit5`.

**Review by mdedetrich — CHANGES_REQUESTED** [review]
While its true that we shouldn't rename the `Junit4` class names, we should actually fix them and have the old class name as an alias that is deprecated (so we don't break binary compatibility).

There is no reason not to improve the current state of things and consistency with current incorrectly named code isn't a compelling argument if you can fix the original issue without breaking users so I would recommend the following

* Rename the classes back to `JUnit4`/`JUnit5` so its actually consistent with how `JUnit` is named.
* Create deprecated `Junit4`/`Junit5` classes that extend `JUnit4`/`JUnit5` so it doesn't break binary compatibility. These can then be dropped in Pekko 2.0.x

File: actor-testkit-typed/src/main/java/org/apache/pekko/actor/testkit/typed/annotations/JUnit5TestKit.java (hunk 2 above)
Comment (mdedetrich) [reply actor-testkit-typed/src/main/java/org/apache/pekko/actor/testkit/typed/annotations/JUnit5TestKit.java]: I disagree, I think the class should be named `JUnit5` and we should have a deprecated alias for `Junit5`.

**Comment by thmue** [comment]
ok i changed all Junit5 names/labels to JUnit5. I suggest to rename the other Junit4 labels / class names in a different PR.

**Comment by mdedetrich** [comment]
@thmue There are some compile errors for Scala 3

File: actor-testkit-typed/src/main/scala/org/apache/pekko/actor/testkit/typed/javadsl/TestKitJUnit5Extension.scala [hunk 3]
```diff
@@ -29,8 +29,8 @@ final class TestKitJUnit5Extension() extends AfterAllCallback with BeforeTestExe
    * Get a reference to the field annotated with @JUnit5Testkit  [[JUnit5TestKit]]
    */
   override def beforeTestExecution(context: ExtensionContext): Unit = {
-
-    context.getTestInstance.ifPresent((instance: AnyRef) => {
+    val testInstance: Option[AnyRef] = Option.when(context.getTestInstance.isPresent)(context.getTestInstance.get())
```
Comment (pjfanning) [inline actor-testkit-typed/src/main/scala/org/apache/pekko/actor/testkit/typed/javadsl/TestKitJUnit5Extension.scala:32]: I think Option.when only appeared in Scala 2.13. We need to support Scala 2.12 still.

File: actor-typed/src/main/scala-jdk-9/org/apache/pekko/actor/typed/internal/jfr/Events.scala [hunk 4]
```diff
@@ -18,7 +18,6 @@ import jdk.jfr.Enabled
 import jdk.jfr.Event
 import jdk.jfr.Label
 import jdk.jfr.StackTrace
-
```
Comment (pjfanning) [inline actor-typed/src/main/scala-jdk-9/org/apache/pekko/actor/typed/internal/jfr/Events.scala:21]: can you undo this?

File: actor-testkit-typed/src/test/resources/application.conf [hunk 5]
```diff
@@ -2,3 +2,4 @@
 
 # used by ActorTestKitSpec
 test.from-application = yes
+test.value = someValue
```
[the remark this answers is not in the document]
Comment (thmue) [reply actor-testkit-typed/src/test/resources/application.conf]: ok done

File: actor-typed/src/main/scala-jdk-9/org/apache/pekko/actor/typed/internal/jfr/Events.scala (hunk 4 above)
Comment (thmue) [reply actor-typed/src/main/scala-jdk-9/org/apache/pekko/actor/typed/internal/jfr/Events.scala]: should now be the same as in main

File: actor-testkit-typed/src/main/scala/org/apache/pekko/actor/testkit/typed/javadsl/TestKitJUnit5Extension.scala (hunk 3 above)
Comment (thmue) [reply actor-testkit-typed/src/main/scala/org/apache/pekko/actor/testkit/typed/javadsl/TestKitJUnit5Extension.scala]: i replaced Option.when with an if statement

**Review by pjfanning — APPROVED** [review]
lgtm - @mdedetrich wdyt? - it may be worth merging as is and worrying about prospective changes to the junit4 existing code later

**Comment by pjfanning** [comment]
@mdedetrich is it possible to merge this?

**Comment by mdedetrich** [comment]
I would like to add the aliases as I mentioned before, if it's tok much work then I can do it in a follow up PR

**Comment by pjfanning** [comment]
> I would like to add the aliases as I mentioned before, if it's tok much work then I can do it in a follow up PR

I think that would be better in a separate PR

**Review by mdedetrich — APPROVED** [review]
lgtm, lets do alias changes in separate PR

**Comment by aseychell** [comment]
I would like to write tests with Junit5 and it seems this PR was merged 8 months ago but this is still not available on any stable release. Are there any plans to make this available please?

**Comment by raboof** [comment]
Good question! I agree we should be working towards an 1.1.0 release. I think the best way to help with that is to look at the issues and PRs linked to the 1.1.0-M2 milestone and see if there's anything in there that you'd feel comfortable picking up.

**Comment by pjfanning** [comment]
> I would like to write tests with Junit5 and it seems this PR was merged 8 months ago but this is still not available on any stable release. Are there any plans to make this available please?

If you don't want to use 1.1.0-M1 (and it is not a full release), you could copy the 4 small classes from this PR into your code base and use them Pekko 1.0.x.

**Pull request merged**